### ISSUE TITLE
torch_geometric.loader clusterData Issue with Partitioning 

### ISSUE THREAD

**Issue by annatn998** [pyg-team/pytorch_geometric#3222]
I'm looping through and trying to cluster large graphs into smaller partitions. When I use the clusterData function like so: 

graph = pytorch_geometric.Data.data object 
graph2 = pytorch_geometric.Data.data object 

clusters1 = ClusterData(graph, 200)
clusters2 = ClusterData(graph2,200)

etc. (I'm doing much more than 2 graphs but it's the same code repeated in a loop) 

The code works for some of the graphs and I get the expected output: 

Computing METIS partitioning... 
Done!

but then half way through I get this error: 

Process finished with exit code -1073741819 (0xC0000005)

I've gotten a similar partitioning error across both a Mac and a windows 10 laptop. If someone could help me out with this or at least point me to some resources where I might go about solving this issue it would be so helpful. 

Environment: 
python=3.7.11 
torch==1.9.0
torch_cluster== 1.5.9
torch_geometric== 2.0.1
torch_scatter== 2.0.8
torch_sparse== 0.6.12
torch_spline_conv== 1.2.1
torchmetrics== 0.5.1

**Comment by rusty1s** [comment]
I think this is due to Metis not handling some of your graphs correctly. Can you check whether the graphs that fail are not directed and do not have isolated nodes?

**Comment by annatn998** [comment]
Yes! I just checked I think you may be right, I commented out all the directed graphs, and that allowed all the other ones to be partitioned. However, I'm unsure about how to determine whether or not there are isolated nodes, as the graph objects are very large (over 100,000 nodes). Is there a way of helping Metis to handle the directed graphs?

**Comment by rusty1s** [comment]
`data.has_isolated_nodes()` should work. However, you already found the issue :) Converting all graphs to undirected ones for computing METIS partitioning should fix your issues.

**Comment by annatn998** [comment]
I will try this!! Does this go into the cluster function directly? I'm a bit afraid to turn them into undirected graphs as the information from the directed edges is important!

**Comment by annatn998** [comment]
It does seem to have some isolated nodes. However, I just checked again, I put in one directed graph with isolated nodes just to make sure, and that one was also partitioned but the others were not, so I don't think it's just being a directed graph anymore or having isolated nodes :(

**Comment by rusty1s** [comment]
I see. You can copy the [`ClusterData`](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/loader/cluster.py#L46-L49) class and apply some modifications to it (adding self-loops, adding reverse edges), e.g.:
```python
edge_index, _ = add_self_loops(data.edge_index)
edge_index, _ = to_undirected(data.edge_index)
adj = SparseTensor(
    row=edge_index[0], col=edge_index[1],
    value=torch.arange(edge_index.size(1), device=data.edge_index.device),    
    sparse_sizes=(N, N))
```

**Comment by annatn998** [comment]
Thank you so much!! This worked for almost every graph :D That really helps I appreciate it! There's just now one smaller graph that seems to fail the partitioning and it looks like this: 

Data(x=[1283, 300], edge_index=[2, 3063], y=[1], num_nodes=1283)

It's very similar to the other directed graphs, so do you know of any other potential blockers I should look at?

**Comment by rusty1s** [comment]
Can you upload one or two examples (`only `edge_index` required) that fail? I can try to look into it.

**Comment by annatn998** [comment]
Yes! Thanks so much for looking into it. I've tried out a number of different parts (10,50,100,200) and found that using the 10 & 50 split will sometimes work but the 100 and 200 fail. Here are some examples of the data objects I'm actually using: 

Data(x=[132862, 300], edge_index=[2, 1730035], y=[1], num_nodes=132862)
Data(x=[138710, 300], edge_index=[2, 1781648], y=[1], num_nodes=138710)
Data(x=[129850, 300], edge_index=[2, 1884510], y=[1], num_nodes=129850)
Data(x=[317793, 300], edge_index=[2, 17421910], y=[1], num_nodes=317793)

Thanks so much again, your help is so appreciated!

**Comment by rusty1s** [comment]
To test, I need the detailed information about `edge_index`. Can you store them and upload them somewhere?
```python
out = [data1.edge_index, data2.edge_index, ...]
torch.save(out, path)
```

**Comment by annatn998** [comment]
I've uploaded the file into a zip file in this repository! Thanks again! 

https://github.com/annatn998/pytorch_output

**Comment by rusty1s** [comment]
This works for me:
```python
import torch
from torch_sparse import SparseTensor

path = './output_edge_index'
edge_indices = torch.load(path)

for edge_index in edge_indices:
    adj = SparseTensor.from_edge_index(edge_index)
    print(adj)
    print(adj.is_symmetric())
    adj = adj.to_symmetric()
    print(adj.is_symmetric())

    for num_parts in [10, 50, 100, 200]:
        print(num_parts)
        adj.partition(num_parts=num_parts)
```

**Comment by diamondspark** [comment]
I'm trying to cluster my large connected graphs. However, my `data.has_isolated_nodes()` returns True and hence Metis fails. However `torch_geometric.utils.contains_isolated_nodes(data.edge_index) `returns False. Please advice.

**Comment by rusty1s** [comment]
The inconsistency between `has_isolated_nodes` and `contains_isolated_nodes` have been already fixed in master, see https://github.com/pyg-team/pytorch_geometric/issues/4138. Regarding your issue, I'm pretty sure you can fix it be running `to_undirected(edge_index)` on your `edge_index` first.

**Comment by diamondspark** [comment]
> @rusty1s Thank you for those pointers. I'm trying to cluster relatively small graphs (About 800 nodes with ~100 node features). For any num_parts>10, it kills my jupyter kernel (segmentation fault). Would you have any pointers on num_parts that should be used or how to solve the memory issue?

This was resolved by ensuring edges are undirected by using `to_undirected(edge_index)`

@rusty1s  Is there an easy way to identify the nodes of the subgraph in each cluster?

**Comment by rusty1s** [comment]
Can you explain in more detail what you are trying to do? Have you looked at [`transforms.LargestConnectedComponents`](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/transforms/largest_connected_components.py)? It looks like a good starting point to look into for what you are trying to achieve.

**Comment by licycle** [comment]
> 
which jupyter kernerl version are you using?
